### DISCUSSION TITLE
AA on lines not working as intended with pushToSprite?

### DISCUSSION THREAD

**Discussion by Zoonie81** [Bodmer/TFT_eSPI#2937] [Q&A - Sprites]
Hi Bodmer!

Many thanks for your hard work with this library. The more I get into it, the more I understand the amount of time you have put in.

I'm playing around with something like a watch face, following bits and parts of your example with the anti-aliased clock. So I'm drawing a line with drawWideLine to a destination coordinate with given angle. Everything works great. The lines are perfectly anti-aliased no matter angle. But as soon I push the sprite containing these lines to another sprite, with pushToSprite for double buffering, the lines become very jaggy. If the sprite is pushed directly with pushSprite, everything looks perfect, except the flickering ofc. Even tried drawWedgeLine to adjust line thickness, but problem remains.

Does the conversion of the drawn line with pushToSprite have known limitations that cause this? Is there a way around this problem?
My project is running on a Nano ESP32, 16-bit colors, byte swap set true if that matters.

Regards,
Z

**Comment by Bodmer** [comment]
That should work. Provide a simple example that demonstrates the problem.

**Reply by Bodmer** [reply]
The test sketch is simple so I had time to create one! This works for me, try it.
```
#include <TFT_eSPI.h>

TFT_eSPI tft = TFT_eSPI();

TFT_eSprite spr1 = TFT_eSprite(&tft);
TFT_eSprite spr2 = TFT_eSprite(&tft);

void setup()   {
  tft.begin();
  tft.setRotation(0);

  spr1.createSprite(50,50);
  spr1.drawWideLine(4, 4, 40, 40, 3, TFT_RED, TFT_BLACK);

 // spr2.setColorDepth(8); // Test pushing to 8 bit (has lower colour resolution)
  spr2.createSprite(50,50);

  spr1.pushToSprite(&spr2,0,0);

  spr1.pushSprite(0,0);
  spr2.pushSprite(50,0);
}

void loop() {
}

```
Note that pushing to sprites of different bit widths might not work so well.

Can you provide and example that does not work?

**Comment by Zoonie81** [comment]
Dear Bodmer! Thanks for taking time!
Since I was somewhat in a hurry and couldn’t find a solution, I ended up rewriting my code completly and so now I don’t have the old ready code with the phenomenon to show you. I was doing something like this though: I made a fullscreen sprite called “canvas”. On this sprite I pushed a fullscreen image from a .h file, like a wallpaper (16-bits, byte swap true). All good. Then I had a temporary sprite that I drew lines on, and then I pushed that onto the canvas with TFT_BLACK set to transparency and finally pushSprite. That’s when the line became jaggy.

Part of the code doing this looked something along these lines.. Nothing complex really.

````
TFT_eSPI LCD = TFT_eSPI();
TFT_eSprite canvas = TFT_eSprite(&LCD);
TFT_eSprite tempSprite = TFT_eSprite(&LCD);

void setup() {
  LCD.init();
  LCD.fillScreen(TFT_BLACK);
  canvas.createSprite(240, 240);
  canvas.setSwapBytes(true);

  tempSprite.createSprite(120, 50);
  tempSprite.setTextColor(TFT_WHITE, TFT_BLACK);
  tempSprite.setTextDatum(MC_DATUM);
}

void loop() {
  canvas.pushImage( 0,0,240,240,Wallpaper);
  tempSprite.drawWedgeLine(x2_ind, y2_ind, x_ind, y_ind, 1, 1, TFT_WHITE, TFT_BLACK);
  tempSprite.pushToSprite(&canvas, 0, 0, TFT_BLACK);
  canvas.pushSprite(0,0);
}
````

I should've given you my actual code in the first post, sorry about that. Anyway, I tried drawWideLine with the same results. Changed thicknesses and same results. Also tried different colors for filling and transparency with same results. I ended up rewriting my algorithm for making the lines and the layout so it could be plotted directly to canvas without the need for a temp sprite to push from. That fixed everything and I'm all happy now :D I see you push both your sprites at the end, even though spr1 was already pushed to spr2. Seeing them being pushed to different coordinates though, I guess it's for comparisons sake? 

/Z
